packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts

```diff
@@ -325,6 +326,57 @@ export const useGridRows = (
     [apiRef, logger],
   );
 
+  const replaceRows = React.useCallback<GridRowApi['replaceRows']>(
+    (firstRowToRender, lastRowToRender, newRows) => {
+      if (props.signature === GridSignature.DataGrid) {
+        throw new Error(
+          [
+            "MUI: You can't replace rows using `apiRef.current.replaceRows` on the DataGrid.",
+            'You need to upgrade to the DataGridPro component to unlock this feature.',
+          ].join('\n'),
+        );
+      }
+
+      const newRowsIds = newRows.map((row) => row.id);
+      const allRows = gridRowIdsSelector(apiRef);
+      const updatedRows = [...allRows];
+      updatedRows.splice(firstRowToRender, lastRowToRender - firstRowToRender, ...newRowsIds);
+
+      const idRowsLookup = gridRowsLookupSelector(apiRef);
+      const idToIdLookup = gridRowsIdToIdLookupSelector(apiRef);
+      const tree = gridRowTreeSelector(apiRef);
+      const updatedIdRowsLookup = { ...idRowsLookup };
+      const updatedIdToIdLookup = { ...idToIdLookup };
+      const updatedTree = { ...tree };
+      const rowIdsToBeReplaced = allRows.slice(firstRowToRender, lastRowToRender);
+
+      rowIdsToBeReplaced.forEach((id) => {
```

Comment:
There's room for optimization here - `replaceRows` currently loops over `newRows` 3 times, we can reduce it to 2 loops:

```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts b/packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts
index 38446fe86..326ecf52d 100644
--- a/packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts
@@ -337,10 +337,8 @@ export const useGridRows = (
         );
       }
 
-      const newRowsIds = newRows.map((row) => row.id);
       const allRows = gridRowIdsSelector(apiRef);
       const updatedRows = [...allRows];
-      updatedRows.splice(firstRowToRender, lastRowToRender - firstRowToRender, ...newRowsIds);
 
       const idRowsLookup = gridRowsLookupSelector(apiRef);
       const idToIdLookup = gridRowsIdToIdLookupSelector(apiRef);
@@ -348,14 +346,17 @@ export const useGridRows = (
       const updatedIdRowsLookup = { ...idRowsLookup };
       const updatedIdToIdLookup = { ...idToIdLookup };
       const updatedTree = { ...tree };
-      const rowIdsToBeReplaced = allRows.slice(firstRowToRender, lastRowToRender);
 
-      rowIdsToBeReplaced.forEach((id) => {
-        delete updatedIdRowsLookup[id];
-        delete updatedIdToIdLookup[id];
-        delete updatedTree[id];
+      newRows.forEach((row, index) => {
+        const rowId = row.id;
+        const [replacedRowId] = updatedRows.splice(firstRowToRender + index, 1, rowId);
+
+        delete updatedIdRowsLookup[replacedRowId];
+        delete updatedIdToIdLookup[replacedRowId];
+        delete updatedTree[replacedRowId];
       });
 
+      // update lookups in separate loops to avoid deleting rows when sorting has changed
       newRows.forEach((row) => {
```

Initially I wanted to reduce it to a single loop, but then realized that it won't work with sorting. So 2 loops are needed here.